src/common/io.ts

```diff
@@ -154,39 +154,54 @@ export class ARIOReadable implements AoARIORead {
     });
   }
 
+  private async computeEpochIndexForTimestamp(
+    timestamp: number,
+  ): Promise<number> {
+    const epochSettings = await this.getEpochSettings();
+    const epochZeroStartTimestamp = epochSettings.epochZeroStartTimestamp;
+    const epochLengthMs = epochSettings.durationMs;
+    return Math.floor((timestamp - epochZeroStartTimestamp) / epochLengthMs);
+  }
+
+  private async epochInputToEpochIndexTag(
+    params?: EpochInput,
+  ): Promise<{ name: string; value: string }> {
+    const epochIndex = (params as { epochIndex?: number })?.epochIndex;
+
+    if (epochIndex !== undefined) {
+      return { name: 'Epoch-Index', value: epochIndex.toString() };
+    }
+    const timestamp = (params as { timestamp?: number })?.timestamp;
+    if (timestamp !== undefined) {
+      return {
+        name: 'Epoch-Index',
+        value: (await this.computeEpochIndexForTimestamp(timestamp)).toString(),
+      };
+    }
+    return { name: 'Epoch-Index', value: '' };
+  }
+  private epochSettings: AoEpochSettings | undefined;
   async getEpochSettings(params?: EpochInput): Promise<AoEpochSettings> {
+    if (this.epochSettings) {
+      return this.epochSettings;
+    }
+
+    const epochInputTag = await this.epochInputToEpochIndexTag(params);
     const allTags = [
       { name: 'Action', value: 'Epoch-Settings' },
-      {
-        name: 'Timestamp',
-        value:
-          (params as { timestamp?: number })?.timestamp?.toString() ??
-          Date.now().toString(),
-      },
-      {
-        name: 'Epoch-Index',
-        value: (params as { epochIndex?: number })?.epochIndex?.toString(),
-      },
+      epochInputTag,
```

Comment:
i don't believe this handler supports epoch-index, we don't store settings state on the epoch

Reply:
[b311ca](https://github.com/ar-io/ar-io-sdk/pull/325/commits/1b311ca4d21c33ca5a6a44b942fb5e5673c09c37)